### PR TITLE
Add float32 support

### PR DESCRIPTION
**Pull request by FanhaiLu1**
This PR add float32 support for pytorch inference. Below are the components in this PR

- Float32 Weight conversion
- Float32 engine 
- Float32 cache 
- Add test for both Float32 and float16

### CODE REVIEW

File: convert_checkpoints.py [hunk 1]
```diff
@@ -229,6 +235,9 @@ def _merge_weights(checkpoints, minimize_memory_footprint):
                 f'Tensors must be identical across shards for {key}'
             )
           state_dict_for_key[key] = tensors[0]
+
+        if enable_float32:
```
Comment (gangji) [inline convert_checkpoints.py:239]: In general, branch prediction should happen at the outer most loop. Putting `if` condition inside is very expensive. Thanks!

File: jetstream_pt/engine.py [hunk 2]
```diff
@@ -507,8 +508,13 @@ def _load_from_safetensors(self, path):
         arr = jax.device_put(f.get_tensor(key), self.sharding_by_name(key))
         assert tuple(model_weights.shape) == tuple(arr.shape), f"key: {key} error: {model_weights.shape} != {arr.shape}"
         weights[key] = arr
+        print(f"key:{key}, value shape:{arr.dtype}")
     freqs_cis = torch_xla2.tensor.t2j(self.pt_model.freqs_cis)
     weights['freqs_cis'] = jax.device_put(freqs_cis, self.replicated)
+
+    # if not self.env.bf16_enable:
```
Comment (gangji) [inline jetstream_pt/engine.py:515]: ?

**Review by gangji — COMMENTED** [review]
Thanks for the great work!

File: convert_checkpoints.py (hunk 1 above)
Comment (FanhaiLu1) [reply convert_checkpoints.py:239]: No, current approach is already most efficient. There are another condition check inside the loop. If put if outside, the code readiness much worse than current one. On the performance side, both OPs are exactly same. For n condition check and 1 iteration of size k element, doesn't what way you write code, the time complexity is O(K*N).

File: jetstream_pt/engine.py (hunk 2 above)
Comment (FanhaiLu1) [reply jetstream_pt/engine.py:515]: Thanks for caching up. There is another approach to do float32 weights: convert them to float32 in loading time with bfloat16 checking points. In the end, I chose current approach use float32 checking points and comment the convert logics. 

Removed.

File: jetstream_pt/engine.py [hunk 3]
```diff
@@ -507,8 +508,10 @@ def _load_from_safetensors(self, path):
         arr = jax.device_put(f.get_tensor(key), self.sharding_by_name(key))
         assert tuple(model_weights.shape) == tuple(arr.shape), f"key: {key} error: {model_weights.shape} != {arr.shape}"
         weights[key] = arr
+        print(f"key:{key}, value shape:{arr.dtype}")
```
Comment (wang2yn84) [inline jetstream_pt/engine.py:511]: nit: Please follow the same pattern as in 529, and only print layer 0 for better user experience.

**Review by wang2yn84 — COMMENTED** [review]
What's the reason why we are adding back f32? We used to have it but removed cuz we run at bf16 by default and f32 performance is really limited.

File: jetstream_pt/engine.py (hunk 3 above)
Comment (FanhaiLu1) [reply jetstream_pt/engine.py]: Thanks, updated.

**Comment by FanhaiLu1** [comment]
> What's the reason why we are adding back f32? We used to have it but removed cuz we run at bf16 by default and f32 performance is really limited.

It's for E2E accuracy measurement. Since bf16 have token diff on second or third, having fp 32 can help measure the accuracy of the serving infra.
